In [1]:
import sys

In [2]:
print(sys.version)

3.12.1 (main, Jul 10 2025, 11:57:50) [GCC 13.3.0]


In [3]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

In [4]:
pd.__version__

'2.3.1'

In [23]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [24]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


In [9]:
df.shape

(9704, 11)

In [10]:
df.fuel_type.nunique()

2

In [11]:
df.isna().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [12]:
df[df["origin"] == 'Asia']["fuel_efficiency_mpg"].max()

np.float64(23.759122836520497)

In [13]:
df.horsepower.median()

np.float64(149.0)

In [14]:
df["horsepower"].mode()

0    152.0
Name: horsepower, dtype: float64

In [15]:
df.horsepower

0       159.0
1        97.0
2        78.0
3         NaN
4       140.0
        ...  
9699    164.0
9700    154.0
9701    138.0
9702    177.0
9703    140.0
Name: horsepower, Length: 9704, dtype: float64

In [16]:
df['horsepower'] = df.horsepower.fillna(152)

149.0
152.0


-  version = 2.0.3
- values = 9704
- types of fuel = 2
- 4 columns with missing valus 
- fuel_efficiency = 23.759122836520497
- Yes, it increased
- 0.5187709081074046

- Select all the cars from Asia
- Select only columns vehicle_weight and model_year
- Select the first 7 values
- Get the underlying NumPy array. Let's call it X.
- Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
- Invert XTX.
- Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
- Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
- What's the sum of all the elements of the result?

In [17]:
# df_asia = 
X = df[df['origin'] == 'Asia'][['vehicle_weight','model_year']][:7].values

In [18]:
XTX = X.T.dot(X)

In [19]:
XTX_inv = np.linalg.inv(XTX)

In [20]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [21]:
w = (XTX_inv.dot(X.T)).dot(y)

In [22]:
w.sum()

np.float64(0.5187709081074016)